src/zenml/integrations/vault/secrets_manager/vault_secrets_manager.py

```diff
@@ -0,0 +1,342 @@
+        verify: Whether to verify the certificate or not.
+        mount_point: The mount point of the secrets manager.
+        namespace: The namespace of the secrets manager.
+    """
+
+    # Class configuration
+    FLAVOR: ClassVar[str] = VAULT_SECRETS_MANAGER_FLAVOR
+    CLIENT: ClassVar[Any] = None
+
+    url: str
+    token: str
+    cert: Optional[str]
+    verify: Optional[str]
+    namespace: Optional[str]
+    mount_point: Optional[str] = "secret"
+
+    @classmethod
+    def _ensure_client_connected(cls, url: str, token: str) -> None:
+
+        if cls.CLIENT is None:
+            # Create a Vault Secrets Manager client
+            cls.CLIENT = hvac.Client(
+                url=url,
+                token=token,
+            )
+
+    def _ensure_client_is_authenticated(self) -> None:
+        """
+        Ensure the client is authenticated.
+
+        Raises:
+            RuntimeError: If the client is not initialized or authenticated.
+        """
+
+        self._ensure_client_connected(url=self.url, token=self.token)
+
+        if not self.CLIENT:
+            raise RuntimeError("Vault client is not initialized.")
+        else:
+            if not self.CLIENT.is_authenticated():
+                raise RuntimeError("Vault client is not authenticated.")
+            else:
+                pass
+
+    def register_secret(self, secret: BaseSecretSchema) -> None:
+        """Registers a new secret.
+
+        Args:
+            secret: The secret to register.
+
+        Raises:
+            SecretExistsError: If the secret already exists.
+        """
+        self._ensure_client_is_authenticated()
+
+        secret_name = prepend_secret_schema_to_secret_name(secret=secret)
+
+        if secret_name in self.get_all_secret_keys():
+            raise SecretExistsError(
+                f"A Secret with the name '{secret_name}' already exists."
+            )
+
+        self.CLIENT.secrets.kv.v2.create_or_update_secret(
+            path=f"{ZENML_PATH}/{secret_name}",
+            secret=secret.content,
+        )
+
+        logger.debug("Created secret: %s", f"{ZENML_PATH}/{secret_name}")
+        logger.debug("Added value to secret.")
+
+    def get_secret(self, secret_name: str) -> BaseSecretSchema:
+        """Gets the value of a secret.
+
+        Args:
+            secret_name: The name of the secret to get.
+
+        Returns:
+            The secret.
+
+        Raises:
+            SecretDoesNotExistError: If the secret does not exist.
+        """
+
+        try:
+            vault_secret_name = self.vault_secret_name(secret_name)
+        except SecretDoesNotExistError:
+            raise SecretDoesNotExistError(
+                f"The secret {secret_name} does not exist."
+            )
+
+        secret_items = (
+            self.CLIENT.secrets.kv.v2.read_secret_version(
+                path=f"{ZENML_PATH}/{vault_secret_name}",
+                mount_point=self.mount_point,
+            )
+            .get("data", {})
```

Comment:
Any reason that this has to be run twice?

Reply:
Thank you @AlexejPenner  for the review, I appreciate the time and effort!! 

You have raised an important point here. However, the reason is related to the way the response is defined by `hvac – Python client for Vault`. As it returns secrets according to the format : 

```json
{
  "request_id": "...", 
  "lease_id": "...", 
  "renewable": "...", 
  "lease_duration": "...", 
  "data": {
    "data": {"..."}, 
    "metadata": {
      "created_time": "...", 
      "custom_metadata": "...", 
      "deletion_time": "...", 
      "destroyed": "...", 
      "version": "..."
      }
  }, 
  "wrap_info": "...", 
  "warnings": "...", 
  "auth": "..."
}
```

Reply:
Totally makes sense.